File: python3/longest_valid_parentheses.py
"""
Space   : O(1)
Time    : O(n)
"""

class Solution:
    def longestValidParentheses(self, s: str) -> int:
        n = len(s)
        ans = 0
        
        # forward
        start, par = 0, 0
        for end in range(n):
            if s[end] == "(":
                par += 1
            else:
                par -= 1
                
            if par < 0:
                start = end + 1
                par = 0
            elif par == 0:
                ans = max(ans, end-start+1)
            
        # backward
        end, par = n-1, 0
        for start in range(n-1, -1, -1):
            if s[start] == "(":
                par += 1
            else:
                par -= 1
                
            if par > 0:
                end = start - 1
                par = 0
            elif par == 0:
                ans = max(ans, end-start+1)
                
        return ans

File: golang/longest_valid_parentheses.go
package main

/*
Space	: O(1)
Time	: O(n)

Method:
two pointer iteration
*/


func longestValidParentheses(s string) int {
	ans, n := 0, len(s)

	// forward
	count, start := 0, 0
	for i := 0; i < n; i++ {
		if string(s[i]) == "(" {
			count++
		} else {
			count--
		}

		if count < 0 {
			count = 0
			start = i + 1
		} else if count == 0 {
			ans = Max(ans, i-start+1)
		}
	}

	// backward
	count, end := 0, n-1
	for i := n - 1; i > -1; i-- {
		if string(s[i]) == "(" {
			count--
		} else {
			count++
		}

		if count < 0 {
			count = 0
			end = i - 1
		} else if count == 0 {
			ans = Max(ans, end-i+1)
		}
	}

	return ans
}
